Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1486,101 +1487,196 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
+          ? DecodeSnapshotVersion(reader, write_result.update_time)
+          : commit_version;
+
+  absl::optional<std::vector<FieldValue>> transform_results;
+  if (write_result.transform_results_count > 0) {
+    transform_results = std::vector<FieldValue>{};
+    for (pb_size_t i = 0; i < write_result.transform_results_count; i++) {
+      transform_results->push_back(
+          DecodeFieldValue(reader, write_result.transform_results[i]));
+    }
+  }
+
+  return MutationResult(version, std::move(transform_results));
 }
 
 std::unordered_map<std::string, std::string>
 Serializer::EncodeListenRequestLabels(const QueryData& query_data) const {
-  // TODO(varconst): implement.
-  // - (nullable NSMutableDictionary<NSString *, NSString *> *)
-  //     encodedListenRequestLabelsForQueryData: (const QueryData &)queryData;
-  return {};
+  auto value = EncodeLabel(query_data.purpose());
+  if (value.empty()) {
+    return {};
+  }
+
+  std::unordered_map<std::string, std::string> result;
+  result["goog-listen-tags"] = std::move(value);
+  return result;
 }
 
 std::string Serializer::EncodeLabel(QueryPurpose purpose) const {
-  // TODO(varconst): implement.
-  // - (nullable NSString *)encodedLabelForPurpose:(QueryPurpose)purpose;
-  return {};
+  switch (purpose) {
+    case QueryPurpose::Listen:
+      return "";
+    case QueryPurpose::ExistenceFilterMismatch:
+      return "existence-filter-mismatch";
+    case QueryPurpose::LimboResolution:
+      return "limbo-document";
+  }
+  UNREACHABLE();
 }
 
 std::unique_ptr<WatchChange> Serializer::DecodeWatchChange(
     nanopb::Reader* reader,
     const google_firestore_v1_ListenResponse& watch_change) const {
-  // TODO(varconst): implement.
-  // - (std::unique_ptr<WatchChange>)decodedWatchChange:
-  //     (GCFSListenResponse *)watchChange;
-  return {};
+  switch (watch_change.which_response_type) {
+    case google_firestore_v1_ListenResponse_target_change_tag:
+      return DecodeTargetChange(reader, watch_change.target_change);
+
+    case google_firestore_v1_ListenResponse_document_change_tag:
+      return DecodeDocumentChange(reader, watch_change.document_change);
+
+    case google_firestore_v1_ListenResponse_document_delete_tag:
+      return DecodeDocumentDelete(reader, watch_change.document_delete);
+
+    case google_firestore_v1_ListenResponse_document_remove_tag:
+      return DecodeDocumentRemove(reader, watch_change.document_remove);
+
+    case google_firestore_v1_ListenResponse_filter_tag:
+      return DecodeExistenceFilterWatchChange(reader, watch_change.filter);
+  }
+  UNREACHABLE();
 }
 
 SnapshotVersion Serializer::DecodeVersion(
     nanopb::Reader* reader,
     const google_firestore_v1_ListenResponse& listen_response) const {
-  // TODO(varconst): implement.
-  // - (SnapshotVersion)versionFromListenResponse:
-  //     (GCFSListenResponse *)watchChange;
-  return {};
+  // We have only reached a consistent snapshot for the entire stream if there
```

Comment:
Can you add a new test for this function?

Reply:
Note: Android calls this `decodeVersionFromListenResponse` and calls the thing that decodes the read time `decodeVersion`. Could you make line up more closely?

Reply:
Done.